src/NHibernate/Engine/ActionQueue.cs

```diff
@@ -56,6 +57,8 @@ public ActionQueue(ISessionImplementor session)
 
 			afterTransactionProcesses = new AfterTransactionCompletionProcessQueue(session);
 			beforeTransactionProcesses = new BeforeTransactionCompletionProcessQueue(session);
+
+			querySpaces = new HashSet<string>();
```

Comment:
I think it is better to utilize `afterTransactionProcesses.querySpacesToInvalidate` instead of creating an additional hashset.

Reply:
Or opposite.

Reply:
Or opposite. As you have 2 collections which hold exactly the same data.

Reply:
Also, you've forgot to clear the field.